torch/_higher_order_ops/associative_scan.py

```diff
@@ -0,0 +1,173 @@
+
+
+def trace_associative_scan(proxy_mode, func_overload, input, dim, combine_fn):
+    pre_dispatch = getattr(proxy_mode, "pre_dispatch", False)
+
+    with disable_proxy_modes_tracing():
+        sample_inputs = (
+            torch.full((), False, dtype=input.dtype, device=input.device),
+            torch.full((), False, dtype=input.dtype, device=input.device),
+        )
+        combine_graph = make_fx(
+            _maybe_run_with_interpreter(combine_fn), pre_dispatch=pre_dispatch
+        )(*sample_inputs)
+
+    outputs = []
+    for node in combine_graph.graph.nodes:
+        if node.op == "output":
+            outputs.extend(node.args)
+
+    torch._check(
+        len(outputs) == 1,
+        lambda: f"expected combine_fn to have 1 output but got {len(outputs)}",
+    )
+
+    for o in outputs:
+        o_meta = o.meta["tensor_meta"]
+        torch._check(
+            o_meta.dtype == input.dtype,
+            lambda: (
+                f"combine_fn output type mismatch, expected {input.dtype} "
+                + f"but got {o_meta.dtype}"
+            ),
+        )
+        torch._check(
+            o_meta.shape == (),
+            lambda: f"combine_fn must return a scalar tensor but got shape {o_meta.shape}",
+        )
+
+    _, combine_graph_name = unique_graph_id(proxy_mode, prefix="scan_combine_graph")
+
+    proxy_mode.tracer.root.register_module(combine_graph_name, combine_graph)
+
+    args = (input, dim, combine_graph)
+    proxy_args = pytree.tree_map(proxy_mode.tracer.unwrap_proxy, args)
+    out_proxy = proxy_mode.tracer.create_proxy(
+        "call_function", func_overload, proxy_args, {}, name="associative_scan"
+    )
+
+    with disable_proxy_modes_tracing():
+        out = aten.clone(input)
+
+    return track_tensor_tree(out, out_proxy, constant=None, tracer=proxy_mode.tracer)
+
+
+@associative_scan_op.py_impl(DispatchKey.CompositeExplicitAutograd)
+def associative_scan_op_dense(input, dim, combine_fn):
+    raise NotImplementedError("associative_scan is not implemented for eager")
+
+
+associative_scan_op.py_impl(DispatchKey.Autograd)(
+    autograd_not_implemented(associative_scan_op, deferred_error=True)
+)
+
+
+@associative_scan_op.py_impl(ProxyTorchDispatchMode)
+def associative_scan_proxy_mode(mode, input, dim, combine_fn):
+    if mode.enable_tracing:
+        return trace_associative_scan(mode, associative_scan_op, input, dim, combine_fn)
+    else:
+        return associative_scan_op(mode, associative_scan_op, input, dim, combine_fn)
+
+
+@associative_scan_op.py_impl(FakeTensorMode)
+def assoiciative_scan_fake_tensor_mode(mode, input, dim, combine_fn):
+    with mode:
+        return input.clone()
+
+
+@associative_scan_op.py_functionalize_impl
+def associative_scan_functionalize(ctx, input, dim, combine_fn):
+    unwrapped_input = ctx.unwrap_tensors(input)
+    with ctx.redispatch_to_next() as m:
+        ret = associative_scan_op(unwrapped_input, dim, combine_fn)
+    return ctx.wrap_tensors(ret)
```

Comment:
Are we saying that the combine_fn must not mutate any tensors?

Reply:
Yes I don't think it makes sense to support mutations inside the combine_fn.